app/models/reducer.rb

```diff
@@ -43,67 +45,127 @@ def self.of_type(type)
   validates :topic, presence: true
   validates_associated :extract_filter
 
-  before_validation :nilify_empty_fields
-
   config_field :user_reducer_keys, default: nil
   config_field :subject_reducer_keys, default: nil
 
   NoData = Class.new
 
-  def process(extract_fetcher, reduction_fetcher, relevant_reductions=[])
+  def process(extracts, reductions, subject_id=nil, user_id=nil)
+    @subject_id = subject_id
+    @user_id = user_id
+
     light = Stoplight("reducer-#{id}") do
-      extract_fetcher.strategy = running_reduction?  ? :fetch_minimal : :fetch_all
-      grouped_extracts = ExtractGrouping.new(extract_fetcher.extracts, grouping).to_h
+      return [] if extracts.empty?
+
+      grouped_extracts = ExtractGrouping.new(extracts, grouping).to_h
+
+      reduced_extracts = grouped_extracts.map do |group_key, extract_group|
+        # find or create the reduction we want to reduce into
+        reduction = get_group_reduction(reductions, group_key)
 
-      grouped_extracts.map do |group_key, grouped|
-        reduction = get_reduction(reduction_fetcher, group_key)
-        extracts = filter_extracts(grouped, reduction)
-        next if extracts.empty?
+        # now that we have the reduction, filter the extracts based
+        # on the filters and on reduction.extracts if it exists
+        extracts = filter_extracts(extract_group, reduction)
 
-        # Set relevant reduction on each extract if required by external reducer
-        # relevant_reductions are any previously reduced user or subject reductions
-        # that are required by this reducer to properly calculate
-        augmented_extracts = add_relevant_reductions(extracts, relevant_reductions)
+        # use the decorator to isolate reducer state in running vs default mode
+        reduction_state = ReductionState.new(reduction, running_reduction?)
```

Comment:
This is the decorator that i've added in, currently this isn't great to test as this needs to be setup each time to inject into the custom reducer `reduce_into` method. 

Help finding a nice interface please...

Reply:
I think for me the thing is that in this file it's not clear why this is needed, and at this point in the code (as in reading top-down) you can't easily predict what the decorator is going to do.

It's also not really a decorator, since `reduce_into` is going to change its state. Perhaps reducers need to be changed so that they always just return a new `{data: foo, store: bar}` hash, or possibly a new ReductionState. Then have this base class take care of updating the reduction with the new returned state.

Reply:
Also note that that style is how reducers originally worked: they would just return the reduced `data`. https://github.com/zooniverse/caesar/pull/263 changed this in commit 2a0dd78. I can't remember anymore exactly why, there was something that was supposedly easier but I can't think of any blockers for going back.